src/wasm/wasm-s-parser.cpp

```diff
@@ -2858,27 +2858,32 @@ void SExpressionWasmBuilder::parseData(Element& s) {
   if (!wasm.memory.exists) {
     throw ParseException("data but no memory", s.line, s.col);
   }
-  bool isPassive = false;
+  bool isPassive = true;
   Expression* offset = nullptr;
   Index i = 1;
   Name name;
-  if (s[i]->dollared()) {
+
+  if (s[i]->isStr() && s[i]->dollared()) {
     name = s[i++]->str();
   }
-  if (s[i]->isStr()) {
-    // data is passive or named
-    if (s[i]->str() == PASSIVE) {
-      isPassive = true;
+
+  if (s[i]->isList()) {
+    // Optional (memory <memoryidx>)
+    if (elementStartsWith(s[i], MEMORY)) {
+      i += 1;
```

Comment:
It would be good to add a TODO here about supporting multiple memories propoerly.

Reply:
Done. Thanks!